heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+heapster_define_llvmshape            heapster_init_env_for_files
+heapster_define_opaque_llvmshape     heapster_init_env_for_files_debug
+heapster_define_opaque_perm          heapster_init_env_from_file
+heapster_define_perm                 heapster_init_env_from_file_debug
+heapster_define_reachability_perm    heapster_join_point_hint
+heapster_define_recursive_perm       heapster_parse_test
+heapster_define_rust_type            heapster_print_fun_trans
+heapster_define_rust_type_qual       heapster_set_debug_level
+heapster_export_coq                  heapster_set_translation_checks
+heapster_find_symbol                 heapster_typecheck_fun
+heapster_find_symbol_commands        heapster_typecheck_fun_rename
+heapster_find_symbol_with_type       heapster_typecheck_mut_funs
+```
+
+You can then use `:?` to see further information for each of them.
+
+
+#### 3. Load the file and extract the two function specifications.
+
+To load a file into heapster you can use `heapster_init_env`. Let's
+check it's documentation first
+
+``` bash
+sawscript> :help heapster_init_env
+Description
+-----------
+
+EXPERIMENTAL
+
+heapster_init_env : String -> String -> TopLevel HeapsterEnv
+
+Create a new Heapster environment with the given SAW module name
+from the named LLVM bitcode file.
+```
+
+As you see it takes two names. The first, is the name of the SAW core
+module where Heapster will write all the specifications we extract. If
+you start with a fresh name, Heapster will autogenerate a new module
+for you. We will see later how to use existisng modules. The second
+name referse to the bitecode file containing the code we are
+verifying.
+
+The function returns a Heapster environment that contains all the
+definitions of the module (not to be confused with the SAW environment
+that can be printed wiht `:env`).
+
+Let's load our module with
+
+```
+env <- heapster_init_env "linked_list" "examples/linked_list.bc";
+```
+
+we have created a new Heapster environment that we can explore. 
+
+```
+sawscript> print env
+[20:19:48.436] Module: examples/linked_list.bc
+Types:
+  %struct.list64_t = type { i64, %struct.list64_t* }
+
+Globals:
+
+External references:
+  declare default void @llvm.dbg.declare(metadata, metadata,
+                                         metadata)
+  declare default i8* @malloc(i64)
+
+Definitions:
+  i64 @is_head(i64 %0, %struct.list64_t* %1)
+  i64 @is_elem(i64 %0, %struct.list64_t* %1)
+  i64 @any(i64(i64)* %0, %struct.list64_t* %1)
+  %struct.list64_t* @find_elem(i64 %0, %struct.list64_t* %1)
+  %struct.list64_t* @sorted_insert(i64 %0, %struct.list64_t* %1)
+  %struct.list64_t* @sorted_insert_no_malloc(%struct.list64_t* %0,
+                                             %struct.list64_t* %1)
+```
+
+The heapster environment contains all the types, global definitions,
+external references and functions from the loaded module.
+
+
+#### 4. Writing heapster specifications for your functions
+
+##### Pen and paper specification
+
```

Comment:
You should use the term "Heapster types" instead of "Heapster specifications", because, like you say later, Heapster types aren't full functional specifications, but instead mostly deal with memory shapes and whether LLVM values are pointers or word values.